File: backend/app/scheduling/rrule_codec.py
"""RFC 5545 RRULE codec and friendly-builder pattern mapping."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, date, datetime

from dateutil.rrule import rrulestr

# Friendly-builder pattern ids (UI / codec only; not stored in DB).
PATTERN_NONE = "none"
PATTERN_WEEKDAYS = "weekdays"
PATTERN_EVERY_N_DAYS = "every_n_days"
PATTERN_EVERY_N_WEEKS = "every_n_weeks"
PATTERN_EVERY_N_MONTHS = "every_n_months"
PATTERN_EVERY_N_YEARS = "every_n_years"
PATTERN_SPECIFIC_DATES = "specific_dates"
PATTERN_MONTH_DAYS = "month_days"

_ISO_TO_ICAL = {
    1: "MO",
    2: "TU",
    3: "WE",
    4: "TH",
    5: "FR",
    6: "SA",
    7: "SU",
}


@dataclass(frozen=True)
class FriendlyPattern:
    pattern: str
    interval: int | None = None
    weekdays: list[int] | None = None
    month_days: list[int] | None = None


def is_recurring(rrule: str | None, rdates: list[date] | None) -> bool:
    if rrule:
        return True
    return bool(rdates)


def validate_rrule(rrule: str) -> None:
    """Raise ValueError when [rrule] is not parseable."""
    rrulestr(f"RRULE:{rrule}", dtstart=datetime(2000, 1, 1, tzinfo=UTC))


def pattern_to_rrule(
    pattern: str,
    *,
    interval: int | None = None,
    weekdays: list[int] | None = None,
    month_days: list[int] | None = None,
    until: datetime | None = None,
) -> str | None:
    if pattern in (PATTERN_NONE, PATTERN_SPECIFIC_DATES):
        return None

    n = interval or 1
    parts: list[str] = []

    if pattern == PATTERN_EVERY_N_DAYS:
        parts = ["FREQ=DAILY", f"INTERVAL={n}"]
    elif pattern == PATTERN_EVERY_N_WEEKS:
        parts = ["FREQ=WEEKLY", f"INTERVAL={n}"]
    elif pattern == PATTERN_WEEKDAYS:
        if not weekdays:
            raise ValueError("weekdays required for weekdays pattern")
        byday = ",".join(_ISO_TO_ICAL[d] for d in sorted(weekdays))
        parts = ["FREQ=WEEKLY", f"INTERVAL={n}", f"BYDAY={byday}"]
    elif pattern == PATTERN_EVERY_N_MONTHS:
        parts = ["FREQ=MONTHLY", f"INTERVAL={n}"]
    elif pattern == PATTERN_EVERY_N_YEARS:
        parts = ["FREQ=YEARLY", f"INTERVAL={n}"]
    elif pattern == PATTERN_MONTH_DAYS:
        if not month_days:
            raise ValueError("month_days required for month_days pattern")
        bymonthday = ",".join(str(d) for d in sorted(month_days))
        parts = ["FREQ=MONTHLY", f"INTERVAL={n}", f"BYMONTHDAY={bymonthday}"]
    else:
        raise ValueError(f"Unknown pattern: {pattern}")

    if until is not None:
        until_utc = until.astimezone(UTC)
        parts.append(f"UNTIL={until_utc.strftime('%Y%m%dT%H%M%SZ')}")

    return ";".join(parts)


def _weekday_to_iso(wd: object) -> int:
    if hasattr(wd, "weekday"):
        return int(wd.weekday) + 1  # type: ignore[attr-defined]
    if isinstance(wd, int):
        return wd + 1
    raise ValueError(f"Unsupported BYDAY value: {wd!r}")


def rrule_to_pattern(rrule: str) -> FriendlyPattern:
    rule = rrulestr(f"RRULE:{rrule}", dtstart=datetime(2000, 1, 1, tzinfo=UTC))
    interval = rule._interval or 1
    freq = rule._freq
    has_byday = "BYDAY=" in rrule.upper()
    has_bymonthday = "BYMONTHDAY=" in rrule.upper()

    if freq == 3:  # DAILY
        return FriendlyPattern(PATTERN_EVERY_N_DAYS, interval=interval)
    if freq == 2:  # WEEKLY
        if rule._byweekday and has_byday:
            iso_days = sorted({_weekday_to_iso(wd) for wd in rule._byweekday})
            return FriendlyPattern(
                PATTERN_WEEKDAYS, interval=interval, weekdays=iso_days
            )
        return FriendlyPattern(PATTERN_EVERY_N_WEEKS, interval=interval)
    if freq == 1:  # MONTHLY
        if rule._bymonthday and has_bymonthday:
            days = sorted(int(d) for d in rule._bymonthday)
            return FriendlyPattern(
                PATTERN_MONTH_DAYS, interval=interval, month_days=days
            )
        return FriendlyPattern(PATTERN_EVERY_N_MONTHS, interval=interval)
    if freq == 0:  # YEARLY
        return FriendlyPattern(PATTERN_EVERY_N_YEARS, interval=interval)

    raise ValueError(f"Unsupported RRULE: {rrule}")

File: frontend/lib/core/scheduling/rrule_codec.dart
import 'package:rrule/rrule.dart';

/// Friendly-builder pattern ids (codec only; not stored in DB).
abstract final class ScheduleRepeatType {
  static const none = 'none';
  static const weekdays = 'weekdays';
  static const everyNDays = 'every_n_days';
  static const everyNWeeks = 'every_n_weeks';
  static const everyNMonths = 'every_n_months';
  static const everyNYears = 'every_n_years';
  static const specificDates = 'specific_dates';
  static const monthDays = 'month_days';
  static const quota = 'quota';
}

const _isoToIcal = {
  1: 'MO',
  2: 'TU',
  3: 'WE',
  4: 'TH',
  5: 'FR',
  6: 'SA',
  7: 'SU',
};

class ScheduleRrulePattern {
  const ScheduleRrulePattern({
    required this.pattern,
    this.interval = 1,
    this.weekdays = const [],
    this.monthDays = const [],
  });

  final String pattern;
  final int interval;
  final List<int> weekdays;
  final List<int> monthDays;
}

bool scheduleIsRecurring(
  String? rrule,
  List<DateTime> rdates, {
  int? quotaTimes,
  int? quotaPeriodWeeks,
}) {
  if (quotaTimes != null && quotaPeriodWeeks != null) return true;
  if (rrule != null && rrule.isNotEmpty) return true;
  return rdates.isNotEmpty;
}

String? patternToRrule({
  required String pattern,
  int interval = 1,
  List<int> weekdays = const [],
  List<int> monthDays = const [],
  DateTime? until,
}) {
  switch (pattern) {
    case ScheduleRepeatType.none:
    case ScheduleRepeatType.specificDates:
    case ScheduleRepeatType.quota:
      return null;
    case ScheduleRepeatType.everyNDays:
      return _withUntil('FREQ=DAILY;INTERVAL=$interval', until);
    case ScheduleRepeatType.everyNWeeks:
      return _withUntil('FREQ=WEEKLY;INTERVAL=$interval', until);
    case ScheduleRepeatType.weekdays:
      if (weekdays.isEmpty) {
        throw ArgumentError('weekdays required');
      }
      final byday = weekdays.map((d) => _isoToIcal[d]!).join(',');
      return _withUntil('FREQ=WEEKLY;INTERVAL=$interval;BYDAY=$byday', until);
    case ScheduleRepeatType.everyNMonths:
      return _withUntil('FREQ=MONTHLY;INTERVAL=$interval', until);
    case ScheduleRepeatType.everyNYears:
      return _withUntil('FREQ=YEARLY;INTERVAL=$interval', until);
    case ScheduleRepeatType.monthDays:
      if (monthDays.isEmpty) {
        throw ArgumentError('monthDays required');
      }
      final days = monthDays.join(',');
      return _withUntil('FREQ=MONTHLY;INTERVAL=$interval;BYMONTHDAY=$days', until);
    default:
      throw ArgumentError('Unknown pattern: $pattern');
  }
}

String _withUntil(String rrule, DateTime? until) {
  if (until == null) return rrule;
  final utc = until.toUtc();
  final stamp =
      '${utc.year.toString().padLeft(4, '0')}${utc.month.toString().padLeft(2, '0')}${utc.day.toString().padLeft(2, '0')}T${utc.hour.toString().padLeft(2, '0')}${utc.minute.toString().padLeft(2, '0')}${utc.second.toString().padLeft(2, '0')}Z';
  return '$rrule;UNTIL=$stamp';
}

ScheduleRrulePattern rruleToPattern(String rrule) {
  final rule = RecurrenceRule.fromString('RRULE:$rrule');
  final interval = rule.interval ?? 1;
  final hasByDay = rrule.toUpperCase().contains('BYDAY=');
  final hasByMonthDay = rrule.toUpperCase().contains('BYMONTHDAY=');

  switch (rule.frequency) {
    case Frequency.daily:
      return ScheduleRrulePattern(
        pattern: ScheduleRepeatType.everyNDays,
        interval: interval,
      );
    case Frequency.weekly:
      if (rule.byWeekDays.isNotEmpty && hasByDay) {
        final weekdays = rule.byWeekDays.map((d) => d.day).toSet().toList()
          ..sort();
        return ScheduleRrulePattern(
          pattern: ScheduleRepeatType.weekdays,
          interval: interval,
          weekdays: weekdays,
        );
      }
      return ScheduleRrulePattern(
        pattern: ScheduleRepeatType.everyNWeeks,
        interval: interval,
      );
    case Frequency.monthly:
      if (rule.byMonthDays.isNotEmpty && hasByMonthDay) {
        final days = rule.byMonthDays.map((d) => d).toList()..sort();
        return ScheduleRrulePattern(
          pattern: ScheduleRepeatType.monthDays,
          interval: interval,
          monthDays: days,
        );
      }
      return ScheduleRrulePattern(
        pattern: ScheduleRepeatType.everyNMonths,
        interval: interval,
      );
    case Frequency.yearly:
      return ScheduleRrulePattern(
        pattern: ScheduleRepeatType.everyNYears,
        interval: interval,
      );
    default:
      throw ArgumentError('Unsupported RRULE: $rrule');
  }
}

String scheduleSummaryFromApi(Map<String, dynamic> schedule) {
  final rrule = schedule['rrule']?.toString();
  if (rrule != null && rrule.isNotEmpty) {
    try {
      final pattern = rruleToPattern(rrule);
      final label = _patternLabel(pattern.pattern);
      return label;
    } catch (_) {
      return 'Repeating';
    }
  }
  final rdates = schedule['rdates'];
  if (rdates is List && rdates.isNotEmpty) {
    return 'Specific dates';
  }
  return 'One-off';
}

String _patternLabel(String pattern) => switch (pattern) {
      ScheduleRepeatType.weekdays => 'Weekly on selected days',
      ScheduleRepeatType.everyNDays => 'Every N days',
      ScheduleRepeatType.everyNWeeks => 'Every N weeks',
      ScheduleRepeatType.everyNMonths => 'Every N months',
      ScheduleRepeatType.everyNYears => 'Every N years',
      ScheduleRepeatType.specificDates => 'Specific dates',
      ScheduleRepeatType.monthDays => 'Days of month',
      _ => 'One-off',
    };
